src/metaprob/generative_functions.cljc

```diff
@@ -25,25 +27,29 @@
         trace-with-name
         (fn trace-with-name [expr name]
           (cond
-            (if-expr? expr)
-            `(if ~(trace-with-name (if-predicate expr) name)
-               ~(trace-with-name (if-then-clause expr) name)
-               ~(trace-with-name (if-else-clause expr) name))
+            (code/if-expr? expr)
+            `(if ~(trace-with-name (code/if-predicate expr) name)
+               ~(trace-with-name (code/if-then-clause expr) name)
+               ~(trace-with-name (code/if-else-clause expr) name))
 
-            (do-expr? expr)
+            (code/do-expr? expr)
             (cons 'do (map #(trace-with-name % name) (rest expr)))
 
             (or (not (seq? expr))
                 (special-symbol? (first expr))
-                (let-expr? expr)
-                (let-traced-expr? expr)
-                (fn-expr? expr)
-                (gen-expr? expr))
+                (code/let-expr? expr)
+                (code/let-traced-expr? expr)
+                (code/fn-expr? expr)
+                (code/gen-expr? expr))
             expr
 
             ;; If a macro, try expanding
-            (not= (macroexpand-1 expr) expr)
-            (recur (macroexpand-1 expr) name)
+            (not= #?(:clj (macroexpand-1 expr)
+                     :cljs (ana/macroexpand-1 &env expr))
+                  expr)
+            (recur #?(:clj (macroexpand-1 expr)
+                      :cljs (ana/macroexpand-1 &env expr))
+                   name)
```

Comment:
How expensive is macro expansion? We could get away with expanding only once here if we bind a symbol to a potential expansion, and use that value for the comparison and the recursive call.